### PR TITLE
handled max sampling_factor

### LINKED ISSUE #689: increase max sampling factor in refinement modules
PR #658 added checks on parameters' values. Therefore, the max value of `sampling_factor` should be higher than one, otherwise it does not make sense to have this parameter. I propose to use `100` as maximum value (high enough to have variability when needed, for example in single structure refinement) and to modify the python code so as to handle possible mistakes when the user has, say, 1k models and sampling_factor = 100 (this should not be allowed).

Default and min values are correct.

### PR DESCRIPTION
**Pull request by mgiulini**
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and that you comply with the following criteria:

- [ ] You have sticked to Python. Please talk to us before adding other programming languages to HADDOCK3
- [ ] Your PR is about CNS
- [ ] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [ ] You structured the code into small functions as much as possible. You can use classes if there is a (state) purpose
- [ ] Your code follows our coding style
- [ ] You wrote tests for the new code
- [ ] `tox` tests pass. *Run `tox` command inside the repository folder*
- [ ] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [ ] PR does not add any dependencies, unless permission granted by the HADDOCK team
- [ ] PR does not break licensing
- [ ] Your PR is about writing documentation for already existing code :fire:
- [ ] Your PR is about writing tests for already existing code :godmode:

---

Closes #689 by changing the max value of `sampling_factor` to 100 and adding a check at the python level such that if the product between the number of models and the `sampling_factor` is > 50 000 the code exits with an error.

I have absolutely no idea about why tests are failing..

### CODE REVIEW

**Comment by rvhonorato** [comment]
Tests were failing because of tox, see https://github.com/haddocking/haddock3/pull/693, please merge from main before merging the pr

File: src/haddock/modules/refinement/emref/__init__.py [hunk 1]
```diff
@@ -47,6 +47,12 @@ def _run(self):
             sampling_factor = 1
         if sampling_factor > 100:
             self.log("[Warning] sampling_factor is larger than 100")
+
+        if len(models_to_refine) * sampling_factor > 50000:
```
Comment (rvhonorato) [inline src/haddock/modules/refinement/emref/__init__.py:51]: out of scope here, but these numbers should be configurable constants instead of hardcoded

**Review by rvhonorato — APPROVED** [review]

File: src/haddock/modules/refinement/emref/__init__.py (hunk 1 above)
Comment (VGPReys) [reply src/haddock/modules/refinement/emref/__init__.py:51]: should this parameter be in the `general` ones ?

**Review by VGPReys — CHANGES_REQUESTED** [review]
I must agree with rodrigo, should we place a kind of `overall_max_models` parameter somewhere ? and setting its default value to 50000, which is a lot already...

**Comment by mgiulini** [comment]
> I must agree with rodrigo, should we place a kind of `overall_max_models` parameter somewhere ? and setting its default value to 50000, which is a lot already...

I actually agree with this...there should be a `max_nmodels` parameter to be placed in the `defaults.yaml`..I propose the following default values for the three refinement modules:

- `max_nmodels`=10000 for flexref and mdref
- `max_nmodels`=50000 for emref

What do you say? Then probably the maximum `sampling_factor` can even be raised to 200 or 400

**Comment by amjjbonvin** [comment]
Fine to have max values, but it is at the end up to the users.

Should be properly documented.

Also what if you hit the max, e.g. after a clustering step and increased sampling factor at flexref? Do you stop or only compute to the max number of models. And in that case how do you select the models?

**Comment by VGPReys** [comment]
> > I must agree with rodrigo, should we place a kind of `overall_max_models` parameter somewhere ? and setting its default value to 50000, which is a lot already...
> 
> I actually agree with this...there should be a `max_nmodels` parameter to be placed in the `defaults.yaml`..I propose the following default values for the three refinement modules:
> 
>     * `max_nmodels`=10000 for flexref and mdref
> 
>     * `max_nmodels`=50000 for emref
> 
> 
> What do you say? Then probably the maximum `sampling_factor` can even be raised to 200 or 400

So any modules within the `modules/refinement/*/default.yaml` should have a `max_nmodels` parameter from now ?

I guess the values are OK.
Still can be risky to have a 10000 `mdref` accessible to any user.
Increasing the `explevel` could help not to have crazy `easy` users doing 10000 `mdref`.

Of note: It is now the second time that I think upper/lower boundaries of some parameters should be related to `explevel`.
But I do not know if with the current implementation this is feasable.
- It could be a kind of a new `guru parameters` with default setting to 0 that sum on top of `easy` one ?
- Should some parameters that require high computation be somehow restricted only using the web interface ?

**Comment by mgiulini** [comment]
> Fine to have max values, but it is at the end up to the users. Should be properly documented. Also what if you hit the max, e.g. after a clustering step and increased sampling factor at flexref? Do you stop or only compute to the max number of models. And in that case how do you select the models?

In the proposed implementation haddock3 is stopped right at the beginning of the refinement module (message is "`Too many models to refine. Please reduce the number of models or decrease the sampling_factor`", will add the current and max number of allowed models to it).

I would not do a model selection here, since 99% of the times this limit is hit will be a user mistake.

**Comment by mgiulini** [comment]
> > > I must agree with rodrigo, should we place a kind of `overall_max_models` parameter somewhere ? and setting its default value to 50000, which is a lot already...
> > 
> > 
> > I actually agree with this...there should be a `max_nmodels` parameter to be placed in the `defaults.yaml`..I propose the following default values for the three refinement modules:
> > ```
> > * `max_nmodels`=10000 for flexref and mdref
> > 
> > * `max_nmodels`=50000 for emref
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What do you say? Then probably the maximum `sampling_factor` can even be raised to 200 or 400
> 
> So any modules within the `modules/refinement/*/default.yaml` should have a `max_nmodels` parameter from now ?

I guess so:)
> 
> I guess the values are OK. Still can be risky to have a 10000 `mdref` accessible to any user. Increasing the `explevel` could help not to have crazy `easy` users doing 10000 `mdref`.

Yes of course, the `sampling_factor` is only accessible to expert users!

> Of note: It is now the second time that I think upper/lower boundaries of some parameters should be related to `explevel`. But I do not know if with the current implementation this is feasable.
> 
> * It could be a kind of a new `guru parameters` with default setting to 0 that sum on top of `easy` one ?
> * Should some parameters that require high computation be somehow restricted only using the web interface ?

I think that the overall HADDOCK idea is to give no flexibility to users below a certain expertise level, and full flexibility to the others..In my opinion implementing expertise-specific thresholds is a bit of an overcomplication..

**Comment by rvhonorato** [comment]
I was suggesting a configuration value, not a module parameter; something in `src/haddock/modules/defaults.yaml`. 

what I mean is that imo validation is not in the scope of a module and should be handled in the workflow composition step instead - anyway out of scope of this PR

**Comment by mgiulini** [comment]
> I was suggesting a configuration value, not a module parameter; something in `src/haddock/modules/defaults.yaml`.
> 
> what I mean is that imo validation is not in the scope of a module and should be handled in the workflow composition step instead - anyway out of scope of this PR

certainly out of the scope of this PR, but good point..the problem is that a priori you cannot know how many models will reach the refinement (thinking of clustering here), so there's no way to check this at the workflow composition step

**Review by VGPReys — DISMISSED** [review]

**Review by VGPReys — APPROVED** [review]

**Pull request merged**